internal/registry/aws/auth.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/google/go-containerregistry/pkg/authn"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/fluxcd/image-reflector-controller/internal/registry"
+)
+
+// ParseImage returns the AWS account ID and region and `true` if
+// the image repository is hosted in AWS's Elastic Container Registry,
+// otherwise empty strings and `false`.
+func ParseImage(image string) (accountId, awsEcrRegion string, ok bool) {
+	registryPartRe := regexp.MustCompile(`([0-9+]*).dkr.ecr.([^/.]*)\.(amazonaws\.com[.cn]*)/([^:]+):?(.*)`)
+	registryParts := registryPartRe.FindAllStringSubmatch(image, -1)
+	if len(registryParts) < 1 {
+		return "", "", false
+	}
+	return registryParts[0][1], registryParts[0][2], true
```

Comment:
May lead to index out of bounds:

```suggestion
	if len(registryParts) < 3 {
		return "", "", false
	}
	return registryParts[0][1], registryParts[0][2], true
```

Reply:
Just tried it and looks like it was correct for ensuring `registryParts` is not zero, but the other indices are slices inside that slice, multi-dimensional. We can add another check for that, to make it safer.

Reply:
Just added one more check:
```diff
-       if len(registryParts) < 3 {
+       if len(registryParts) < 1 || len(registryParts[0]) < 3 {
```
Which would make sure it fails and not panics.